it/src/main/java/com/google/cloud/teleport/it/spanner/DefaultSpannerResourceManager.java

```diff
@@ -0,0 +1,219 @@
+      }
+    }
+  }
+
+  @Override
+  public synchronized void insertTableRecord(Mutation tableRecord) throws IllegalStateException {
+    insertTableRecords(ImmutableList.of(tableRecord));
+  }
+
+  @Override
+  public synchronized void insertTableRecords(Iterable<Mutation> tableRecords)
+      throws IllegalStateException {
+    if (spanner.isClosed()) {
+      throw new IllegalStateException(
+          "Manager has cleaned up all resources and cannot complete operation.");
+    }
+    databaseClient.write(tableRecords);
+  }
+
+  @Override
+  public synchronized ResultSet readTableRecords(String tableId, String... columnNames)
+      throws IllegalStateException {
+    return readTableRecords(tableId, ImmutableList.copyOf(columnNames));
+  }
+
+  @Override
+  public synchronized ResultSet readTableRecords(String tableId, Iterable<String> columnNames)
+      throws IllegalStateException {
+    if (spanner.isClosed()) {
+      throw new IllegalStateException(
+          "Manager has cleaned up all resources and cannot complete operation.");
+    }
+    return databaseClient.singleUse().read(tableId, KeySet.all(), columnNames);
+  }
+
+  @Override
+  public synchronized void cleanupAll() {
+    instanceAdminClient.deleteInstance(instanceId.getInstance());
+    spanner.close();
+  }
+
+  /** Builder for {@link DefaultSpannerResourceManager}. */
+  public static final class Builder {
+    private static final int MAX_BASE_ID_LENGTH = 30;
+    private static final HashFunction BASE_ID_HASH_FUNCTION = goodFastHash(32);
+
+    private final String testId;
+    private final String projectId;
+    private final String region;
+    private String instanceId;
+    private String databaseId;
+
+    private Builder(String testId, String projectId, String region) {
+      this.testId = testId;
+      this.projectId = projectId;
+      this.region = region;
+    }
+
+    public DefaultSpannerResourceManager build() throws CreateInstanceException {
+      String baseIdString = testId.toLowerCase();
+
+      // Make hashing part a function, so I can test?
+      if (baseIdString.length() > MAX_BASE_ID_LENGTH) {
+        String hash = BASE_ID_HASH_FUNCTION.hashUnencodedChars(baseIdString).toString();
+        baseIdString =
+            baseIdString.substring(0, MAX_BASE_ID_LENGTH - hash.length() - 1) + "-" + hash;
+      }
+
+      DateTimeFormatter timeFormat = DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss-SSSSSS");
+      LocalDateTime localDateTime = LocalDateTime.now(ZoneId.of("UTC"));
+      instanceId = baseIdString.replace("_", "-") + "-" + localDateTime.format(timeFormat);
+
+      // replace hyphen with underscore, so there's no need for backticks, then remove trailing
+      // underscores.
+      databaseId = CharMatcher.is('_').trimTrailingFrom(baseIdString.replace("-", "_"));
+
+      DefaultSpannerResourceManager manager = new DefaultSpannerResourceManager(this);
+      manager.createInstance(projectId, region);
```

Comment:
I think the instance needs to be created lazily, similar to the database, only when someone calls `createTable` or `readTableRecords` forst time.